man/fmt.3
.TH qlibs: fmt 3
.SH NAME
fmt \- ASCII representation of strings and integers
.SH SYNTAX
.B #include \(dqfmt.h\(dq

unsigned int \fBfmt_str\fP(char *\fIdest\fR,const char *\fIsource\fR);
.br
unsigned int \fBfmt_strn\fP(char *\fIdest\fR,const char *\fIsource\fR,unsigned int \fImaxlen\fR);

unsigned int \fBfmt_uint\fP(char *\fIdest\fR,unsigned int \fIsource\fR);
.br
unsigned int \fBfmt_uint0\fP(char *\fIdest\fR,unsigned int \fIsource\fR,unsigned int \fIn\fR);
.br
unsigned int \fBfmt_ulong\fP(char *\fIdest\fR,unsigned long \fIsource\fR);
.br
unsigned int \fBfmt_xlong\fP(char *\fIdest\fR,unsigned long \fIsource\fR);
.br 
char \fBtohex\fP(char \fInum\fR);
.br
int \fBfromhex\fP(unsigned char \fIc\fR);
.SH DESCRIPTION
.B fmt_str 
copies all leading nonzero bytes from \fIsource\fR to \fIdest\fR
and returns the number of bytes it copied.
.B fmt_str 
does not append \\0.

.B fmt_strn
copies at most \fImaxlen\fR leading nonzero bytes from
\fIsource\fR to \fIdest\fR and returns the number of bytes it copied.
.B fmt_strn 
does not append \\0.

.B fmt_uint 
writes an ASCII representation ('0' to '9', base 10) of
\fIsource\fR to \fIdest\fR and returns the number of bytes written.
.B fmt_uint 
does not append \\0.

.B fmt_uint0 
writes an ASCII representation ('0' to '9', base 10) of
\fIsource\fR to \fIdest\fR and returns the number of bytes written.
The output is padded with '0'-bytes until it encompasses at least
\fIn\fR bytes, but it will not be truncated if it does not fit.
.B fmt_uint0 
does not append \\0.

.B fmt_ulong 
writes an ASCII representation ('0' to '9', base 10) of
\fIsource\fR to \fIdest\fR and returns the number of bytes written
perhaps including a trailing \\0.
.B fmt_ulong 
does not append \\0.

.B fmt_xlong 
writes an ASCII representation ('0' to '9' and 'a' to 'f', base 16) 
of \fIsource\fR to \fIdest\fR and returns the number of bytes
written.
.B fmt_xlong 
does not append \\0.

.B tohex
reads the ASCII representation of a decimal \fInum\fR and returns its 
hexadecimal ASCII value; thus \'0\' -> \'0\' ... \'9\' -> \'9\',  
\'10\' -> \'a\' and finally \'15\' -> f'.

.B fromhex
reads the ACSII representation of a hexadecimal number \'0\' to \'f\' 
irrelevant of its case and returns its integer value. 

For convenience, 
.B fmt.h 
defines the integers
.I FMT_LEN
and
.I FMT_ULONG
to be big enough to the number of bytes it would have written.


.SH "RETURN CODES"
If \fIdest\fR equals FMT_LEN (i.e. is zero), all
.B fmt_* 
routins return the number of bytes it would have written
i.e. the number of leading nonzero bytes of \fIsource\fR 
perhaps including a \\0.
.SH "SEE ALSO"
byte(3), 
case(3), 
scan(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
	DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
	DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.


fmt.c
#include "fmt.h"

/**
	@file fmt.c
	@author djb
	@source qmail
	@brief formating differnt inputs format for output printing
*/

unsigned int fmt_str(register char *s,register char *t)
{
  register unsigned int len;
  char ch;
  len = 0;
  if (s) { while ((ch = t[len])) s[len++] = ch; }
  else while (t[len]) len++;
  return len;
}

unsigned int fmt_strn(register char *s,register char *t,register unsigned int n)
{
  register unsigned int len;
  char ch;
  len = 0;
  if (s) { while (n-- && (ch = t[len])) s[len++] = ch; }
  else while (n-- && t[len]) len++;
  return len;
}

unsigned int fmt_uint(register char *s,register unsigned int u)
{
  register unsigned long l; l = u; return fmt_ulong(s,l);
}

unsigned int fmt_uint0(char *s,unsigned int u,unsigned int n)
{
  unsigned int len;
  len = fmt_uint(FMT_LEN,u);
  while (len < n) { if (s) *s++ = '0'; ++len; }
  if (s) fmt_uint(s,u);
  return len;
}

unsigned int fmt_ulong(register char *s,register unsigned long u)
{
  register unsigned int len; register unsigned long q;
  len = 1; q = u;
  while (q > 9) { ++len; q /= 10; }
  if (s) {
    s += len;
    do { *--s = '0' + (u % 10); u /= 10; } while(u); /* handles u == 0 */
  }
  return len;
}

unsigned int fmt_xlong(register char *s,register unsigned long u)
{
  register unsigned int len; register unsigned long q;
  len = 1; q = u;
  while (q > 15) { ++len; q /= 16; }
  if (s) {
    s += len;
    do { *--s = tohex(u % 16); u /= 16; } while(u); /* handles u == 0 */
  }
  return len;
}

char tohex(char num) {
  if (num < 10)
    return num + '0';
  else if (num < 16)
    return num - 10 + 'a';
  else
    return -1;
}

int fromhex(unsigned char c) {
  if (c >= '0' && c <= '9')
    return c-'0';
  else if (c >= 'A' && c <= 'F')
    return c -'A' + 10;
  else if (c >= 'a' && c <= 'f')
    return c - 'a' + 10;
  return -1;
}
